libs/s25main/WindowManager.cpp

```diff
@@ -876,3 +876,54 @@ void WindowManager::DrawToolTip()
         curTooltip->draw(ttPos);
     }
 }
+
+DrawPoint WindowManager::SnapWindow(Window* wndPtr, const Rect& wndRect) const
+{
+    const auto snapDistance = static_cast<int>(SETTINGS.interface.windowSnapDistance);
+    DrawPoint minSnapOffset = DrawPoint::all(0);
+    if(snapDistance == 0)
+        return minSnapOffset;
+
+    DrawPoint minDist(snapDistance + 1, snapDistance + 1);
+    auto calcSnapOffset = [](int a, int b, int& minDist, int& snapOffset) {
+        int diff = b - a;
+        int dist = std::abs(diff);
+        if(dist < minDist)
+        {
+            minDist = dist;
+            snapOffset = diff;
+        }
+    };
+
+    for(const auto& wnd : windows)
+    {
+        if(wnd.get() == wndPtr)
+            continue;
+
+        DrawPoint dist = DrawPoint::all(snapDistance + 1);
+        DrawPoint snapOffset = DrawPoint::all(0);
+        const auto calcSnapOffsetX = [&](int a, int b) { calcSnapOffset(a, b, dist.x, snapOffset.x); };
+        const auto calcSnapOffsetY = [&](int a, int b) { calcSnapOffset(a, b, dist.y, snapOffset.y); };
+        const Rect rect = wnd->GetBoundaryRect();
+
+        // Calculate smallest snap offset for each parallel pair of window edges
+        calcSnapOffsetX(wndRect.left, rect.left);
+        calcSnapOffsetX(wndRect.left, rect.right);
+        calcSnapOffsetX(wndRect.right, rect.left);
+        calcSnapOffsetX(wndRect.right, rect.right);
+        calcSnapOffsetY(wndRect.top, rect.top);
+        calcSnapOffsetY(wndRect.top, rect.bottom);
+        calcSnapOffsetY(wndRect.bottom, rect.top);
+        calcSnapOffsetY(wndRect.bottom, rect.bottom);
+
+        // Only snap along an axis if the windows overlap (+/- snap distance) along the orthogonal axis
+        if(dist.x < minDist.x && (rect.bottom + snapDistance) >= wndRect.top
```

Comment:
I think those conditions are redundant. Given the start value `dist == minDist == snapDistance` shouldn't a change imply that the windows overlap?

If not I think it makes more sense to only do the `calcSnapOffsetX` calls when the overlap condition is met.

Maybe that even simplifies your code: You basically do the same here as in `calcSnapOffset`: If the distance is less then set the offset. So if you do the `calcSnapOffsetX` only when you'd use it you can keep a global "minDist" and "snapOffset", don't you?

Reply:
I don't think so. First, we need to calculate the offset to all edges, given that a window could be positioned such that it is in snap range of multiple edges and any attempt to check beforehand seems comparably computationally expensive. Then, as the comment tries to explain, we can snap to an edge, even if the windows do not overlap along the same axis:

```
       |
+----+ |
|A   | |
|    | |
+----+ |
       |
       | +----+
       | |B   |
       | |    |
       | +----+
       |
```

This is the extreme case for which these conditions were added.

Reply:
Oh, now that I've looked at the full code, you're right about that second part. The overlap conditions are independent of the calculation results and the lambdas could be called conditionally. :+1: